Volume expansion

phillc opened the pull request:
Superseeds #29

phillc commented:
I pushed a canary image to docker hub

I created an lke cluster

I applied this manifest

```
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: pvc-test
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 10Gi
  storageClassName: linode-block-storage
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx-deployment
  labels:
    app: nginx
spec:
  replicas: 1
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - name: nginx
        image: nginx:1.14.2
        ports:
        - containerPort: 80
        volumeMounts:
        - mountPath: /testmount
          name: pvc-test
      volumes:
      - name: pvc-test
        persistentVolumeClaim:
          claimName: pvc-test
```

I ran `kubectl apply -f ./e2e/test/manifest/linode-blockstorage-csi-driver.yaml` (the manifest generated in this repo that is pointed at canary

I ran `kubectl edit pvc` and changed the spec's storage to 11gb (from 10)

I now see 11gb in cloud manager.

`kubectl describe pvc` still only shows 10gb ☹️ 

Didn't work but getting closer...

```Warning  VolumeResizeFailed     6m56s                   external-resizer linodebs.csi.linode.com                                              update capacity of PV pvc-f1a09e43e6ed4f95 failed: persistentvolumes "pvc-f1a09e43e6ed4f95" is forbidden: User "system:serviceaccount:kube-system:csi-controller-sa" cannot patch resource "persistentvolumes" in API group "" at the cluster scope
  Warning  VolumeResizeFailed     6m43s (x11 over 6m56s)  external-resizer linodebs.csi.linode.com                                              resize volume pvc-f1a09e43e6ed4f95 failed: rpc error: code = Internal desc = [400] Storage volumes can only be resized up
```

phillc commented:
I pushed a canary image to docker hub

I created an lke cluster

I applied this manifest:

```
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: pvc-test
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 10Gi
  storageClassName: linode-block-storage
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx-deployment
  labels:
    app: nginx
spec:
  replicas: 1
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - name: nginx
        image: nginx:1.14.2
        ports:
        - containerPort: 80
        volumeMounts:
        - mountPath: /testmount
          name: pvc-test
      volumes:
      - name: pvc-test
        persistentVolumeClaim:
          claimName: pvc-test
```

I ran `kubectl apply -f ./e2e/test/manifest/linode-blockstorage-csi-driver.yaml` (the manifest generated in this repo that is pointed at the canary image)

I ran `kubectl edit pvc` and changed the spec's storage to 11gb (from 10)

I now see 11gb in cloud manager.

`kubectl describe pvc` now shows 11gb

```
kubectl describe pvc
Name:          pvc-test
Namespace:     default
StorageClass:  linode-block-storage
Status:        Bound
Volume:        pvc-ad8c93749e744b57
Labels:        <none>
Annotations:   pv.kubernetes.io/bind-completed: yes
               pv.kubernetes.io/bound-by-controller: yes
               volume.beta.kubernetes.io/storage-provisioner: linodebs.csi.linode.com
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:      11Gi
Access Modes:  RWO
VolumeMode:    Filesystem
Mounted By:    nginx-deployment-5dcf9ccb57-tsdkq
Events:
  Type     Reason                  Age                   From                                                                                  Message
  ----     ------                  ----                  ----                                                                                  -------
  Normal   Provisioning            114s                  linodebs.csi.linode.com_csi-linode-controller-0_ec6ecbe2-e718-11ea-8fb2-7e2bd7d1d8e2  External provisioner is provisioning volume for claim "default/pvc-test"
  Normal   ExternalProvisioning    106s (x12 over 4m7s)  persistentvolume-controller                                                           waiting for a volume to be created, either by external provisioner "linodebs.csi.linode.com" or manually created by system administrator
  Normal   Provisioning            95s                   linodebs.csi.linode.com_csi-linode-controller-0_f83b2b52-e718-11ea-ae56-ea9c6371ab33  External provisioner is provisioning volume for claim "default/pvc-test"
  Warning  ExternalExpanding       38s                   volume_expand                                                                         Ignoring the PVC: didn't find a plugin capable of expanding the volume; waiting for an external controller to process this PVC.
  Normal   Resizing                38s                   external-resizer linodebs.csi.linode.com                                              External resizer is resizing volume pvc-ad8c93749e744b57
  Normal   VolumeResizeSuccessful  34s                   external-resizer linodebs.csi.linode.com                                              Resize volume succeeded
```

asauber commented on e2e/test/manifest/linode-blockstorage-csi-driver.yaml:
```diff
@@ -0,0 +1,560 @@
+# pkg/linode-bs/deploy/kubernetes/01-csi-nodeinfo.yaml
```
Based on `test` depending upon `test-manifest` Makefile, this file should not be committed, and it should be gitignored in this directory. Maybe you committed it by mistake?

phillc replied:
It was committed in the original PR, so I left it. I'll remove it though.

asauber commented on pkg/linode-bs/controllerserver.go:
```diff
@@ -348,19 +348,20 @@ func (linodeCS *LinodeControllerServer) ListVolumes(ctx context.Context, req *cs
 	}
 
+	if req.GetMaxEntries() > 0 {
+		nextToken = strconv.Itoa(int(req.GetMaxEntries()))
```
`nextToken` should always be set to a Linode page number, as written this would skip entire pages of volumes on subsequent invocations

This function is invoked by CSI multiple times: first with a "starting token" which takes a provider-specific pagination offset, then again with the "NextToken" returned by the previous invocation. (This scheme allows for both numerical paging and identity based paging that you might see with something like Kafka. Linode uses numerical paging.)

phillc replied:
I've confirmed your assessment. It appears the changes to this file around pagination is in response to csi-test being updated in the original PR, which tested things like respecting max entries. It appears to correctly react to the specified max entries, we must use it as page_size in our API, and then nextToken set in the lines above (to be the next page) would suffice.

However, linodego does not currently support specifying page_size

asauber commented on pkg/linode-bs/driver.go:
```diff
@@ -175,6 +181,8 @@ func (linodeDriver *LinodeDriver) Run(endpoint string) {
 		glog.V(4).Infof("BS Volume Prefix: %v", linodeDriver.bsPrefix)
 	}
 
+	linodeDriver.ready = true
```
The mutex `readyMu` is added, but it's not used to protect the write here.

asauber commented on pkg/linode-bs/identityserver.go:
```diff
@@ -58,11 +59,25 @@ func (linodeIdentity *LinodeIdentityServer) GetPluginCapabilities(ctx context.Co
+
+	return &csi.ProbeResponse{
+		Ready: &wrappers.BoolValue{
+			Value: true,
```
`readyMu` is locked, but this `Value:` doesn't return a `LinodeDriver.ready`

phillc commented on pkg/linode-bs/controllerserver.go:
```diff
@@ -349,15 +349,16 @@ func (linodeCS *LinodeControllerServer) ListVolumes(ctx context.Context, req *cs
-				return nil, status.Error(codes.Aborted, fmt.Sprintf("Invalid starting token %v", startingToken))
-			}
-			listOpts.Page = int(startingPage)
+		listOpts.PageSize = int(req.GetMaxEntries())
```
On that note, the minimum is 25. I looked around trying to figure out what actually sets MaxEntries. Where does that value come from? Regardless, I think it may be best to leave that for the api to complain about, rather than have a magic 500.

asauber replied:
CSI services that make calls to this service set MaxEntires: for example, the `external-attacher`:

https://github.com/kubernetes-csi/external-attacher/blob/2d2ab15c96d6b22e9fb7c290b22bafa8d2cee4e1/pkg/attacher/lister.go#L46

In this case MaxEntires has the zero-value of 0

asauber replied:
Leaving the error cases up to the Linode API doesn't seem ideal to me, but there are cases that would require a lot of extra logic. For example, if MaxEntires was 10, then we would need "overlapping" API requests between iterations of this function.

phillc commented on pkg/linode-bs/controllerserver.go:
```diff
@@ -391,6 +391,9 @@ func (linodeCS *LinodeControllerServer) ListVolumes(ctx context.Context, req *cs
 			},
 		})
+		if len(entries) >= int(req.MaxEntries) {
+			break
```
The initial PR added this

```
        if len(entries) == int(req.MaxEntries) {
            nextToken = strconv.Itoa(int(req.MaxEntries))
            break
        }
```
I tried to pull it out, but then csi-test started complaining. I think I'll repurpose this to handle a situation where the request max entries is less than the linode api minimum page size

phillc replied:
I'm just getting rid of the break

phillc commented on pkg/linode-bs/controllerserver.go:
```diff
@@ -450,6 +454,57 @@ func (linodeCS *LinodeControllerServer) ListSnapshots(ctx context.Context, req *
+
+	if vol, err = linodeCS.CloudProvider.GetVolume(ctx, volumeID); err != nil {
+		if apiErr, ok := err.(*linodego.Error); ok && apiErr.Code == 404 {
+			return &csi.ControllerExpandVolumeResponse{}, nil
```
It is strange. The code looks very intentional, but I can't quite figure out why. I cannot find any scenarios from other csi that do this, so I am going to remove the guard.

phillc commented:
I pushed a canary image to docker hub

I created an lke cluster with two nodes

I applied this manifest:

```
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: pvc-test
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 10Gi
  storageClassName: linode-block-storage
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx-deployment
  labels:
    app: nginx
spec:
  replicas: 1
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - name: nginx
        image: nginx:1.14.2
        ports:
        - containerPort: 80
        volumeMounts:
        - mountPath: /testmount
          name: pvc-test
      volumes:
      - name: pvc-test
        persistentVolumeClaim:
          claimName: pvc-test
```

I ran `kubectl edit pvc` and changed the spec's storage to 11gb (from 10)

Saw error

```
error: persistentvolumeclaims "pvc-test" could not be patched: persistentvolumeclaims "pvc-test" is forbidden: only dynamically provisioned pvc can be resized and the storageclass that provisions the pvc must support resize
```

I ran kubectl apply -f ./e2e/test/manifest/linode-blockstorage-csi-driver.yaml (the manifest generated in this repo that is pointed at the canary image)

I ran `kubectl edit pvc` and changed the spec's storage to 11gb (from 10), and it was successful.

I now see 11gb in cloud manager.

`kubectl describe pvc` now shows 11gb

```
  Warning  ExternalExpanding       46s    volume_expand                                                                         Ignoring the PVC: didn't find a plugin capable of expanding the volume; waiting for an external controller to process this PVC.
  Normal   Resizing                38s    external-resizer linodebs.csi.linode.com                                              External resizer is resizing volume pvc-xxxxx
  Normal   VolumeResizeSuccessful  35s    external-resizer linodebs.csi.linode.com                                              Resize volume succeeded
```
